crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,501 @@
+use std::cell::RefCell;
+
+use crate::{module::map_block_type, mutators::OperatorAndByteOffset};
+use wasm_encoder::{Function, Instruction};
+use wasmparser::{Operator, Range, TypeOrFuncType};
+pub struct ASTBuilder;
+
+#[derive(Debug, Default)]
+pub struct AST {
+    // Index of the root node
+    root: usize,
+    // Nodes of the tree
+    nodes: Vec<Node>,
+    // indexes of ifs nodes
+    ifs: Vec<usize>,
+    // indexeds of loop nodes
+    loops: Vec<usize>,
+    // indexes of block nodes
+    blocks: Vec<usize>,
+}
+
+/// Encodes the Wasm AST
+pub trait ASTWriter {
+    /// Encodes a loop node
+    /// Redefine this in case of a mutator
+    ///
```

Comment:
In general, doc comments should have a short overview sentence first (you've done well on this point) and then more details in later paragraphs with empty lines separating paragraphs (this should be fixed in all of these doc comments here).

For example:

```suggestion
    /// Encodes a loop node.
    ///
    /// Redefine this if your implementation mutates loops. For example, if your
    /// implementation unrolls the first iteration of a particular loop, override
    /// this method. For the other loops in the Wasm that your implementation is not
    /// mutating, you can call `write_loop_default`.
```